### ISSUE TITLE
pickle of certain objects fail on builtins with recurse=True

### ISSUE THREAD

**Issue by mmckerns** [uqfoundation/dill#120]
Another very odd corner case… looks like something with missing references to `builtins`

```
Python 3.4.3 (default, May 25 2015, 18:57:37) 
[GCC 4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.40)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import dill
>>> dill.settings['recurse'] = True
>>> import sys
>>> import test_nested as module
>>> module.a = 1234
>>> pik_mod = dill.dumps(module)
>>> module.a = 0
>>> del sys.modules[module.__name__]
>>> del module
>>> module = dill.loads(pik_mod)
>>> module.a 
1234
>>> del sys.modules[module.__name__]
>>> del module
>>> 
>>> import test_mixins as module
>>> module.a = 1234
>>> pik_mod = dill.dumps(module)
>>> module.a = 0
>>> del sys.modules[module.__name__]
>>> del module
>>> module = dill.loads(pik_mod)
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'vars'",), <traceback object at 0x10ea637c8>)
<function _get_attr at 0x10dd38ae8> (<module 'test_mixins' from '/Users/mmckerns/dev/svn/pathos/dill/tests/test_mixins.py'>, 'vars')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 209, in loads
    return load(file)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 199, in load
    obj = pik.load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1038, in load
    dispatch[key[0]](self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1384, in load_reduce
    value = func(*args)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 661, in _get_attr
    return getattr(self, name)
AttributeError: 'module' object has no attribute 'vars'
```

**Comment by mmckerns** [comment]
The above works with most other modules, so it seems to be directly coming from `test_mixins`.
```
Python 3.4.3 (default, May 25 2015, 18:57:37) 
[GCC 4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.40)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import dill
>>> dill.settings['recurse'] = True
>>> def quad(a=1, b=1, c=0):
...   inverted = [False]
...   def invert():
...     inverted[0] = not inverted[0]
...   def dec(f):
...     def func(*args, **kwds):
...       x = f(*args, **kwds)
...       if inverted[0]: x = -x
...       return a*x**2 + b*x + c
...     func.__wrapped__ = f
...     func.invert = invert
...     return func
...   return dec
... 
>>> @quad(a=0,b=2)
... def double_add(*args):
...   return sum(args)
... 
>>> fx = sum([1,2,3])
>>> assert double_add(1,2,3) == 2*fx
>>> double_add.invert()
>>> assert double_add(1,2,3) == -2*fx
>>> _d = dill.copy(double_add)
(<class 'AttributeError'>, AttributeError("'module' object has no attribute 'sum'",), <traceback object at 0x1053cdb48>)
<function _get_attr at 0x1046c9ae8> (<module '__main__' (<_frozen_importlib.SourceFileLoader object at 0x104331c88>)>, 'sum')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 159, in copy
    return loads(dumps(obj, *args, **kwds))
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 209, in loads
    return load(file)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 199, in load
    obj = pik.load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1038, in load
    dispatch[key[0]](self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/pickle.py", line 1384, in load_reduce
    value = func(*args)
  File "/Users/mmckerns/lib/python3.4/site-packages/dill-0.2.4.dev0-py3.4.egg/dill/dill.py", line 661, in _get_attr
    return getattr(self, name)
AttributeError: 'module' object has no attribute 'sum'
```

**Comment by mmckerns** [comment]
```
>>> dill.detect.trace(True)
>>> 
>>> _d = dill.dumps(double_add)
F1: <function quad.<locals>.dec.<locals>.func at 0x1053c01e0>
F2: <function _create_function at 0x1046c9268>
Co: <code object func at 0x104542300, file "<stdin>", line 6>
F2: <function _unmarshal at 0x1046c90d0>
D2: <dict object at 0x1053cd9c8>
Ce: <cell at 0x1042cd4c8: int object at 0x104029e70>
F2: <function _create_cell at 0x1046c9598>
Ce: <cell at 0x1044cc7f8: int object at 0x104029eb0>
Ce: <cell at 0x1053988e8: int object at 0x104029e70>
Ce: <cell at 0x105398b28: function object at 0x1053c0158>
F1: <function double_add at 0x1053c0158>
Co: <code object double_add at 0x1053b9540, file "<stdin>", line 1>
D2: <dict object at 0x1053d3548>
B1: <built-in function sum>
F2: <function _get_attr at 0x1046c9ae8>
M2: <module '__main__' (<_frozen_importlib.SourceFileLoader object at 0x104331c88>)>
F2: <function _import_module at 0x1046c9bf8>
D2: <dict object at 0x1053cd188>
Ce: <cell at 0x105398ac8: list object at 0x1046a5bc8>
D2: <dict object at 0x10453cd48>
F1: <function quad.<locals>.invert at 0x1053c0048>
Co: <code object invert at 0x10451c660, file "<stdin>", line 3>
D2: <dict object at 0x1053d3588>
D2: <dict object at 0x1053d0208>
>>> _d
b'\x80\x03cdill.dill\n_create_function\nq\x00(cdill.dill\n_unmarshal\nq\x01C\xb2\xe3\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x03\x00\x00\x00\x1f\x00\x00\x00s;\x00\x00\x00\x88\x03\x00|\x00\x00|\x01\x00\x8e\x00\x00}\x02\x00\x88\x04\x00d\x01\x00\x19r#\x00|\x02\x00\x0b}\x02\x00n\x00\x00\x88\x00\x00|\x02\x00d\x02\x00\x13\x14\x88\x01\x00|\x02\x00\x14\x17\x88\x02\x00\x17S)\x03N\xe9\x00\x00\x00\x00\xe9\x02\x00\x00\x00\xa9\x00)\x03\xda\x04args\xda\x04kwds\xda\x01x)\x05\xda\x01a\xda\x01b\xda\x01c\xda\x01f\xda\x08invertedr\x03\x00\x00\x00\xfa\x07<stdin>\xda\x04func\x06\x00\x00\x00s\x08\x00\x00\x00\x00\x01\x0f\x01\n\x00\n\x01q\x02\x85q\x03Rq\x04}q\x05X\x04\x00\x00\x00funcq\x06N(cdill.dill\n_create_cell\nq\x07K\x00\x85q\x08Rq\th\x07K\x02\x85q\nRq\x0bh\x07K\x00\x85q\x0cRq\rh\x07h\x00(h\x01C]\xe3\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00G\x00\x00\x00s\n\x00\x00\x00t\x00\x00|\x00\x00\x83\x01\x00S)\x01N)\x01\xda\x03sum)\x01\xda\x04args\xa9\x00r\x03\x00\x00\x00\xfa\x07<stdin>\xda\ndouble_add\x01\x00\x00\x00s\x02\x00\x00\x00\x00\x02q\x0e\x85q\x0fRq\x10}q\x11X\x03\x00\x00\x00sumq\x12cdill.dill\n_get_attr\nq\x13cdill.dill\n_import_module\nq\x14X\x08\x00\x00\x00__main__q\x15\x85q\x16Rq\x17X\x03\x00\x00\x00sumq\x18\x86q\x19Rq\x1asX\n\x00\x00\x00double_addq\x1bNN}q\x1ctq\x1dRq\x1e\x85q\x1fRq h\x07]q!\x88a\x85q"Rq#tq$}q%(X\x06\x00\x00\x00invertq&h\x00(h\x01Ci\xe3\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x13\x00\x00\x00s\x13\x00\x00\x00\x88\x00\x00d\x01\x00\x19\x0c\x88\x00\x00d\x01\x00<d\x00\x00S)\x02N\xe9\x00\x00\x00\x00\xa9\x00r\x02\x00\x00\x00)\x01\xda\x08invertedr\x02\x00\x00\x00\xfa\x07<stdin>\xda\x06invert\x03\x00\x00\x00s\x02\x00\x00\x00\x00\x01q\'\x85q(Rq)}q*h&Nh#\x85q+}q,tq-Rq.X\x0b\x00\x00\x00__wrapped__q/h\x1eutq0Rq1.'
>>> 
```

**Comment by mmckerns** [comment]
Potentially the issue is that is `globalvars` does not pick up  `sum`:
```
>>> dill.detect.globalvars(double_add, recurse=True, builtin=True)
{}
```

**Comment by matsjoyce** [comment]
Does `globalvars` search `double_sum.__closure__`, as `double_sum.__closure__[3].cell_contents` is the original `double_sum` and:
```
>>> dill.detect.globalvars(double_add.__closure__[3].cell_contents, recurse=True, builtin=True)
{'sum': <built-in function sum>}
```

**Comment by mmckerns** [comment]
@matsjoyce: I have been working with a similar solution, but I think yours may be slightly better.

**Comment by mmckerns** [comment]
It wasn't actually the root issue, but I found it.

**Comment by mmckerns** [comment]
fixed by ffad7f6d833d04ad2eaae18f28277fafcfd210eb
